# Predicting MLB Player Salaries: A Batting Performance Analysis
---
# Raw Datsets exploration, cleaning and joins


## Contents:
1. [Introduction](#introduction)
2. [Basic Data Exploration and cleaning](#basic-data-exploration)
    - 2.1 [WAR Dataset](#war-dataset)
    - 2.2 [Batting Dataset](#batting-dataset)
    - 2.3 [People Dataset](#people-dataset)
    - 2.4 [Salary Dataset](#salary-dataset)
    - 2.5 [Teams Dataset](#teams-dataset)
3. [Table Joins](#table-joins)
4. [Next Steps](#summary)

---

## 1. Introduction

This notebook provides a basic first exploration of several datasets related to baseball statistics. The datasets cover a wide range of information, including player performance, team performance, player biographical information and salaries. The goal of this exploration is to gain a better understanding of the structure and content of these datasets, perform some basic cleaning and preparation to ensure the data is well-suited for subsequent analysis. And finally, we're going to merge these datasets into a single, comprehensive one for further exploration and analysis.


In [371]:
# Import necessary libraries
import pandas as pd
import numpy as np

#### Dataset Loading <a class="anchor" id="dataset-loading"></a>

In [372]:
# Load datasets
raw_war = pd.read_csv('Raw datasets/war_daily_bat.csv')
raw_teams = pd.read_csv('Raw datasets/Teams.csv')
raw_batting = pd.read_csv('Raw datasets/Batting.csv')
raw_fielding = pd.read_csv('Raw datasets/Fielding.csv')
raw_people = pd.read_csv('Raw datasets/People.csv')
raw_salaries = pd.read_csv('Raw datasets/salary_history.csv')

---
## 2. Basic Data Exploration and Cleaning  <a class="anchor" id="basic-data-exploration"></a>  
We'll start by examining the shape and features of each dataset, and then perform some basic cleaning and preparation to ensure the data is well-suited for subsequent analysis. This includes converting data types where necessary, and dropping irrelevant columns. We'll also address any inconsistencies or anomalies we come across during our exploration.



### 2.1 WAR (Wins Above Replacement) Dataset <a class="anchor" id="war-dataset"></a>

In [373]:
pd.set_option('display.max_columns', None)
# Shape and first rows
print(f'raw_war shape: {raw_war.shape}')
print(f'Rows: {raw_war.shape[0]}')
print(f'Columns: {raw_war.shape[1]}')
raw_war.head()

raw_war shape: (121375, 48)
Rows: 121375
Columns: 48


name_common   age    mlb_ID  player_ID  year_ID team_ID  stint_ID lg_ID  \
0  David Aardsma  22.0  430911.0  aardsda01     2004     SFG         1    NL   
1  David Aardsma  24.0  430911.0  aardsda01     2006     CHC         1    NL   
2  David Aardsma  25.0  430911.0  aardsda01     2007     CHW         1    AL   
3  David Aardsma  26.0  430911.0  aardsda01     2008     BOS         1    AL   
4  David Aardsma  27.0  430911.0  aardsda01     2009     SEA         1    AL   

    PA   G   Inn  runs_bat  runs_br  runs_dp  runs_field  runs_infield  \
0  0.0  11  10.7      0.00      0.0      0.0         0.0           0.0   
1  3.0  43  53.0     -0.90      0.0      0.0         0.0           0.0   
2  0.0   2  32.3      0.00      0.0      0.0         0.0           0.0   
3  1.0   5  48.7     -0.29      0.0      0.0         0.0           0.0   
4  0.0   3  71.3      0.00      0.0      0.0         0.0           0.0   

   runs_outfield  runs_catcher  runs_defense  runs_position  runs_position_p  \
0            0.0           0.0           0.0           0.00             0.00   
1            0.0           0.0           0.0           0.01             0.46   
2            0.0           0.0           0.0           0.00             0.00   
3            0.0           0.0           0.0           0.00             0.14   
4            0.0           0.0           0.0           0.00             0.00   

   runs_replacement  runs_above_rep  runs_above_avg  runs_above_avg_off  \
0               0.0             0.0             0.0                 0.0   
1               0.0            -0.4            -0.4                -0.4   
2               0.0             0.0             0.0                 0.0   
3               0.0            -0.2            -0.2                -0.2   
4               0.0             0.0             0.0                 0.0   

   runs_above_avg_def   WAA  WAA_off  WAA_def   WAR  WAR_def  WAR_off  \
0                 0.0  0.00     0.00    -0.01  0.00    -0.01     0.00   
1                 0.0 -0.04    -0.04    -0.01 -0.04    -0.01    -0.04   
2                 0.0  0.00     0.00     0.00  0.00     0.00     0.00   
3                 0.0 -0.02    -0.02     0.00 -0.02     0.00    -0.02   
4                 0.0  0.00     0.00     0.00  0.00     0.00     0.00   

   WAR_rep    salary pitcher  teamRpG   oppRpG  oppRpPA_rep  oppRpG_rep  \
0      0.0  300000.0       Y  4.67092  4.67092      0.08651     4.67092   
1      0.0       NaN       Y  4.85675  4.86675      0.09085     4.86457   
2      0.0  387500.0       Y  4.85895  4.85895      0.08422     4.85895   
3      0.0  403250.0       Y  4.67400  4.70400      0.08092     4.69650   
4      0.0  419000.0       Y  4.79788  4.79788      0.08302     4.79788   

   pyth_exponent  pyth_exponent_rep  waa_win_perc  waa_win_perc_off  \
0          1.890              1.890         0.500             0.500   
1          1.912              1.913         0.499             0.499   
2          1.912              1.912         0.500             0.500   
3          1.893              1.894         0.497             0.497   
4          1.905              1.905         0.500             0.500   

   waa_win_perc_def  waa_win_perc_rep  OPS_plus  TOB_lg  TB_lg  
0               0.5            0.5000       NaN   0.000  0.000  
1               0.5            0.4998    -100.0   0.694  0.896  
2               0.5            0.5000       NaN   0.000  0.000  
3               0.5            0.4992    -100.0   0.345  0.434  
4               0.5            0.5000       NaN   0.000  0.000

##### Dictionary

| Column Name | Description |  
| --- | --- |
| name_common | Player name |
| age | Player age |
| mlb_ID | MLB ID |
| player_ID | Player ID |
| year_ID | Year (Season played) |
| team_ID | ID of the team they played for the season |
| stint_ID | The stint of the player in that year (a player can have more than one stint in a year if they moved teams). |
| lg_ID | The league the player played in |
| PA | Plate appearances |
| G | Games played |
| Inn | Innings played |
| runs_bat | Runs above average |
| runs_br | Runs from baserunning |
| runs_dp | Runs from avoiding double plays |
| runs_field | Runs from fielding |
| runs_infield | Runs from infield |
| runs_outfield | Runs from outfield |
| runs_catcher | Runs from catching |
| runs_defense | Runs from defense |
| runs_position | Runs from position |
| runs_position_p | Runs from position as pitcher |
| runs_replacement | Runs from replacement |
| runs_above_rep | Runs above replacement |
| runs_above_avg | Runs above average |
| runs_above_avg_off | Runs above average as batter |  
| runs_above_avg_def | Runs above average as fielder |
| WAA | Wins above average |
| WAA_off | Wins above average as batter |
| WAA_def | Wins above average as fielder |
| WAR | Wins above replacement |
| WAR_def | Wins above replacement as fielder |
| WAR_off | Wins above replacement as batter |
| WAR_rep | Wins above replacement as replacement |
| salary | Player salary |
| pitcher | Whether the player is a pitcher |
| teamRpG | Team runs per game |
| oppRpG | Opponent runs per game |
| oppRpPA_rep | Opponent runs per plate appearance as replacement |
| oppRpG_rep | Opponent runs per game as replacement |
| pyth_exponent | Pythagorean exponent |
| pyth_exponent_rep | Pythagorean exponent as replacement |
| waa_win_perc | Win percentage above average |
| waa_win_perc_off | Win percentage above average as batter |
| waa_win_perc_def | Win percentage above average as fielder |
| waa_win_perc_rep | Win percentage above average as replacement |
| OPS_plus | OPS+ |
| TOB_lg | Times on base in league |
| TB_lg | Total bases in league |

The WAR dataset provides a comprehensive view of player performance, capturing the value of a player in all facets of the game including batting, baserunning, fielding, and pitching. The dataset consists of 121,375 rows and 48 columns.

The dataset also includes a variety of other metrics related to player performance. The goal of the WAR metric, in particular, is to summarize a player's total contributions to their team in one statistic.


In [374]:
# Data types
raw_war.dtypes

name_common            object
age                   float64
mlb_ID                float64
player_ID              object
year_ID                 int64
team_ID                object
stint_ID                int64
lg_ID                  object
PA                    float64
G                       int64
Inn                   float64
runs_bat              float64
runs_br               float64
runs_dp               float64
runs_field            float64
runs_infield          float64
runs_outfield         float64
runs_catcher          float64
runs_defense          float64
runs_position         float64
runs_position_p       float64
runs_replacement      float64
runs_above_rep        float64
runs_above_avg        float64
runs_above_avg_off    float64
runs_above_avg_def    float64
WAA                   float64
WAA_off               float64
WAA_def               float64
WAR                   float64
WAR_def               float64
WAR_off               float64
WAR_rep               float64
salary    

It looks that all the data types are correct, so we can move on to the next step.


The majority of these columns are complex and difficult to interpret and they not add much value to our analysis. Therefore, we will drop a subset of these columns to simplify the dataset and make it easier to work with. We will keep the columns that are most relevant to our analysis.

In [375]:
cols_to_keep = ['name_common',
 'age',
 'mlb_ID',
 'player_ID',
 'year_ID',
 'team_ID',
 'stint_ID',
 'lg_ID',
 'PA',
 'G',
 'Inn', 'salary', 'OPS_plus', 'WAR', 'WAR_def', 'WAR_off', 'pitcher']

The columns we are keeping include basic player information (like name, age, and team), performance metrics that are easy to interpret (like Plate Appearances (`PA`), Games played (`G`), and Innings played (`Inn`), and metrics that are likely to be relevant to salary (like Wins Above Replacement (`WAR`), defensive and offensive WAR, and whether the player is a pitcher). We're also keeping the `OPS_plus` column, which is a more advanced metric but is widely used and understood in baseball analytics.

In [376]:
# Drop columns
raw_war = raw_war[cols_to_keep]

In [378]:
# Drop rows with year = 2023
raw_war = raw_war[raw_war['year_ID'] != 2023]

Some pre cleaning

In [379]:
# Remove special characters and punctuation from name_common
raw_war['name_common'] = raw_war['name_common'].str.replace(r"[\"\',.]", '')

C:\Users\Hector G. Guerrero\AppData\Local\Temp\ipykernel_29804\2269050005.py:2: FutureWarning: The default value of regex will change from True to False in a future version.
  raw_war['name_common'] = raw_war['name_common'].str.replace(r"[\"\',.]", '')


In [380]:
raw_war.shape

(120757, 17)

##### `stint_ID` column
In our dataset, a player can have multiple entries for a single season due to having multiple stints. This can occur when a player is traded or moves teams during the season. Each stint is recorded as a separate row, which is not ideal for our analysis as we want a single row per player per season.

To resolve this, we'll aggregate the data at the player and season level. This means we'll combine the statistics for all stints a player had in a single season.

In [381]:
# New dataframe to preserve the first team the player played for in a season
raw_war_teams = raw_war[['player_ID', 'year_ID', 'team_ID', 'stint_ID']]

# drop rows with stint_ID > 1
raw_war_teams = raw_war_teams[raw_war_teams['stint_ID'] == 1]

# Drop stint_ID column
raw_war_teams.drop('stint_ID', axis=1, inplace=True)

In [382]:
# group raw_war 
raw_war_grouped = raw_war.groupby(['name_common', 'age', 'year_ID', 'player_ID', 'pitcher']).sum().reset_index()
raw_war_grouped.shape

C:\Users\Hector G. Guerrero\AppData\Local\Temp\ipykernel_29804\4071232131.py:2: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  raw_war_grouped = raw_war.groupby(['name_common', 'age', 'year_ID', 'player_ID', 'pitcher']).sum().reset_index()


(109262, 15)

In [383]:
# Join raw_war_grouped with df_war_teams
war_pre = raw_war_grouped.merge(raw_war_teams, on=['player_ID', 'year_ID'])

In [384]:
# Change stint_ID 
war_pre['stint_ID'] = np.where(war_pre['stint_ID'] == 1, 1, 
                                        np.where(war_pre['stint_ID'] == 3, 2, 
                                            np.where(war_pre['stint_ID'] == 6, 3, 
                                                np.where(war_pre['stint_ID'] == 10, 4, 
                                                    np.where(war_pre['stint_ID'] == 15, 5, 0)))))

In [385]:
# Sanity check, check for players with more than one team in a year
# Joey Gallo
war_pre[war_pre['name_common'] == 'Joey Gallo']

name_common   age  year_ID  player_ID pitcher     mlb_ID  stint_ID  \
56527  Joey Gallo  21.0     2015  gallojo01       N   608336.0         1   
56528  Joey Gallo  22.0     2016  gallojo01       N   608336.0         1   
56529  Joey Gallo  23.0     2017  gallojo01       N   608336.0         1   
56530  Joey Gallo  24.0     2018  gallojo01       N   608336.0         1   
56531  Joey Gallo  25.0     2019  gallojo01       N   608336.0         1   
56532  Joey Gallo  26.0     2020  gallojo01       N   608336.0         1   
56533  Joey Gallo  27.0     2021  gallojo01       N  1216672.0         2   
56534  Joey Gallo  28.0     2022  gallojo01       N  1216672.0         2   

          PA    G     Inn      salary    OPS_plus   WAR  WAR_def  WAR_off  \
56527  123.0   36   264.0         0.0   91.332137  0.32     0.07     0.21   
56528   30.0   17    72.0         0.0   -2.863830 -0.52    -0.24    -0.32   
56529  532.0  145  1208.0    537120.0  118.426407  2.92    -0.63     3.32   
56530  577.0  148  1199.0    560000.0  108.778664  2.41    -0.01     1.86   
56531  297.0   70   614.3    605500.0  144.925708  3.08     0.47     2.54   
56532  226.0   57   475.7   4400000.0   86.573416  1.51     1.11     0.20   
56533  616.0  153  1304.4   6200000.0  230.520541  4.61     0.92     3.04   
56534  410.0  126   944.4  10275000.0  162.570976  0.23    -0.13    -0.18   

      team_ID  
56527     TEX  
56528     TEX  
56529     TEX  
56530     TEX  
56531     TEX  
56532     TEX  
56533     TEX  
56534     NYY

In [386]:
# Shape
print(f'raw_war shape: {raw_war.shape}')

raw_war shape: (120757, 17)


In [387]:
# Save to csv
war_pre.to_csv('pre_datasets\war_pre.csv', index=False)

### 2.2 Batting Dataset <a class="anchor" id="batting-dataset"></a>



In [388]:
# Shape and first rows
print(f'raw_batting shape: {raw_batting.shape}')
print(f'Rows: {raw_batting.shape[0]}')
print(f'Columns: {raw_batting.shape[1]}')
raw_batting.sample(5)

raw_batting shape: (112184, 22)
Rows: 112184
Columns: 22


playerID  yearID  stint teamID lgID    G   AB   R   H  2B  3B  HR  \
11441   schreos01    1908      2    CHA   AL    6   16   1   3   0   0   0   
90365   corcoro01    2008      1    SEA   AL   50    1   0   0   0   0   0   
103447  hernaen02    2017      1    LAN   NL  140  297  46  64  24   2  11   
40290    kingch01    1959      2    SLN   NL    5    7   0   3   0   0   0   
110278  suzukku01    2021      1    LAA   AL   72  219  17  49   8   0   6   

         RBI   SB   CS  BB    SO  IBB   HBP   SH   SF  GIDP  
11441    0.0  0.0  NaN   1   3.0  NaN   0.0  2.0  NaN   NaN  
90365    0.0  0.0  0.0   0   1.0  0.0   0.0  0.0  0.0   0.0  
103447  37.0  3.0  0.0  41  80.0  2.0   0.0  1.0  3.0   4.0  
40290    1.0  0.0  0.0   0   2.0  0.0   0.0  0.0  1.0   0.0  
110278  16.0  0.0  0.0  12  44.0  0.0  11.0  2.0  3.0   7.0

##### Dictionary

| Column Name | Description |
| --- | --- |
| playerID | Player ID |
| yearID | Year (Season played) |
| stint | The stint of the player in that year (a player can have more than one stint in a year if they moved teams). |
| teamID | ID of the team they played for the season |
| lgID | The league the player played in |
| G | Games played |
| AB | At bats |
| R | Runs |
| H | Hits |
| 2B | Doubles |
| 3B | Triples |
| HR | Homeruns |
| RBI | Runs batted in |
| SB | Stolen bases |
| CS | Caught stealing |
| BB | Walks |
| SO | Strikeouts |
| IBB | Intentional walks |
| HBP | Hit by pitch |
| SH | Sacrifice hits |
| SF | Sacrifice flies |
| GIDP | Grounded into double plays |



The batting dataset is a comprehensive collection of player batting statistics. It contains 112,184 rows and 22 columns, each row representing a player's performance in a particular season.

The dataset includes both pitchers and position players. Pitchers typically have fewer at-bats and different offensive statistics than position players, which is something to keep in mind during the analysis.

In [389]:
# Data types
raw_batting.dtypes

playerID     object
yearID        int64
stint         int64
teamID       object
lgID         object
G             int64
AB            int64
R             int64
H             int64
2B            int64
3B            int64
HR            int64
RBI         float64
SB          float64
CS          float64
BB            int64
SO          float64
IBB         float64
HBP         float64
SH          float64
SF          float64
GIDP        float64
dtype: object

It looks that all the data types are correct, so we can move on to the next step.

For now, we will keep all the columns in the batting dataset. We can always drop unnecessary columns later once we understand the data better.

##### `stint` column
Same situation as in the WAR dataset, a player can have multiple entries for a single season due to having multiple stints. This can occur when a player is traded or moves teams during the season. Each stint is recorded as a separate row, which is not ideal for our analysis as we want a single row per player per season.

To resolve this, we'll follow the same process as we did for the WAR dataset.

In [390]:
# New dataframe 
raw_batting_teams = raw_batting[['playerID', 'yearID', 'teamID', 'stint']]
raw_batting_teams.shape

(112184, 4)

In [391]:
# drop rows with stint > 1
raw_batting_teams = raw_batting_teams[raw_batting_teams['stint'] == 1]

# drop stint column
raw_batting_teams.drop('stint', axis=1, inplace=True)

In [392]:
# group raw_batting 
raw_batting_grouped = raw_batting.groupby(['playerID', 'yearID']).sum().reset_index()

C:\Users\Hector G. Guerrero\AppData\Local\Temp\ipykernel_29804\456594340.py:2: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  raw_batting_grouped = raw_batting.groupby(['playerID', 'yearID']).sum().reset_index()


In [393]:
# Join raw_batting_grouped with raw_batting_teams 
batting_pre = raw_batting_grouped.merge(raw_batting_teams, on=['playerID', 'yearID'])

In [394]:
# Sanity check, check for players with more than one team in a year
# Joey Gallo
batting_pre[batting_pre['playerID'] == 'gallojo01']

playerID  yearID  stint    G   AB   R    H  2B  3B  HR   RBI   SB  \
31674  gallojo01    2015      1   36  108  16   22   3   1   6  14.0  3.0   
31675  gallojo01    2016      1   17   25   2    1   0   0   1   1.0  1.0   
31676  gallojo01    2017      1  145  449  85   94  18   3  41  80.0  7.0   
31677  gallojo01    2018      1  148  500  82  103  24   1  40  92.0  3.0   
31678  gallojo01    2019      1   70  241  54   61  15   1  22  49.0  4.0   
31679  gallojo01    2020      1   57  193  23   35   8   0  10  26.0  2.0   
31680  gallojo01    2021      3  153  498  90   99  13   1  38  77.0  6.0   
31681  gallojo01    2022      3  126  350  48   56   8   2  19  47.0  3.0   

        CS   BB     SO  IBB  HBP   SH   SF  GIDP teamID  
31674  0.0   15   57.0  3.0  0.0  0.0  0.0   0.0    TEX  
31675  0.0    5   19.0  0.0  0.0  0.0  0.0   0.0    TEX  
31676  2.0   75  196.0  1.0  8.0  0.0  0.0   3.0    TEX  
31677  4.0   74  207.0  4.0  3.0  0.0  0.0   3.0    TEX  
31678  2.0   52  114.0  4.0  2.0  1.0  1.0   0.0    TEX  
31679  0.0   29   79.0  2.0  4.0  0.0  0.0   0.0    TEX  
31680  0.0  111  213.0  5.0  6.0  0.0  1.0   6.0    TEX  
31681  0.0   56  163.0  0.0  3.0  0.0  1.0   0.0    NYA

In [395]:
# drop stint column
batting_pre.drop('stint', axis=1, inplace=True)

In [396]:
# Save to csv
batting_pre.to_csv('pre_datasets/batting_pre.csv', index=False)

### 2.3 People Dataset <a class="anchor" id="people-dataset"></a>



In [397]:
# Shape and first rows
print(f'raw_people shape: {raw_people.shape}')
print(f'Rows: {raw_people.shape[0]}')
print(f'Columns: {raw_people.shape[1]}')
raw_people.head()


raw_people shape: (20811, 24)
Rows: 20811
Columns: 24


playerID  birthYear  birthMonth  birthDay birthCountry birthState  \
0  aardsda01     1981.0        12.0      27.0          USA         CO   
1  aaronha01     1934.0         2.0       5.0          USA         AL   
2  aaronto01     1939.0         8.0       5.0          USA         AL   
3   aasedo01     1954.0         9.0       8.0          USA         CA   
4   abadan01     1972.0         8.0      25.0          USA         FL   

    birthCity  deathYear  deathMonth  deathDay deathCountry deathState  \
0      Denver        NaN         NaN       NaN          NaN        NaN   
1      Mobile     2021.0         1.0      22.0          USA         GA   
2      Mobile     1984.0         8.0      16.0          USA         GA   
3      Orange        NaN         NaN       NaN          NaN        NaN   
4  Palm Beach        NaN         NaN       NaN          NaN        NaN   

  deathCity nameFirst nameLast       nameGiven  weight  height bats throws  \
0       NaN     David  Aardsma     David Allan   215.0    75.0    R      R   
1   Atlanta      Hank    Aaron     Henry Louis   180.0    72.0    R      R   
2   Atlanta    Tommie    Aaron      Tommie Lee   190.0    75.0    R      R   
3       NaN       Don     Aase  Donald William   190.0    75.0    R      R   
4       NaN      Andy     Abad   Fausto Andres   184.0    73.0    L      L   

        debut   finalGame   retroID    bbrefID  
0  2004-04-06  2015-08-23  aardd001  aardsda01  
1  1954-04-13  1976-10-03  aaroh101  aaronha01  
2  1962-04-10  1971-09-26  aarot101  aaronto01  
3  1977-07-26  1990-10-03  aased001   aasedo01  
4  2001-09-10  2006-04-13  abada001   abadan01

##### Dictionary

| Column Name | Description |
| --- | --- |
| playerID | Player ID |
| birthYear | Year of birth |
| birthMonth | Month of birth |
| birthDay | Day of birth |
| birthCountry | Country of birth |
| birthState | State of birth |
| birthCity | City of birth |
| deathYear | Year of death |
| deathMonth | Month of death |
| deathDay | Day of death |
| deathCountry | Country of death |
| deathState | State of death |
| deathCity | City of death |
| nameFirst | First name |
| nameLast | Last name |
| nameGiven | Full name |
| weight | Weight in pounds |
| height | Height in inches |
| bats | Batting hand (left, right, or both) |
| throws | Throwing hand (left or right) |
| debut | Date of MLB debut |
| finalGame | Date of final MLB game |
| retroID | Retro ID |
| bbrefID | Baseball Reference ID |


The people dataset contains biographical information about baseball players. This dataset has 20,811 rows and 24 columns. Each row represents a player, and the columns provide various details about the player, such as their birth and death details, physical attributes, and career details.

This dataset will be useful for adding context to our analysis, such as the player's age during each season, their physical attributes, and their career span.

In [398]:
# Data types
raw_people.dtypes

playerID         object
birthYear       float64
birthMonth      float64
birthDay        float64
birthCountry     object
birthState       object
birthCity        object
deathYear       float64
deathMonth      float64
deathDay        float64
deathCountry     object
deathState       object
deathCity        object
nameFirst        object
nameLast         object
nameGiven        object
weight          float64
height          float64
bats             object
throws           object
debut            object
finalGame        object
retroID          object
bbrefID          object
dtype: object

It looks that all the data types are correct, so we can move on to the next step.

##### Personal columns

Some columns in this dataset are not relevant to our project and will be dropped. These columns are commonly related to personal information about the player, such as birth date, birth place, and death date. 

Personal columns are not directly related to the project's analysis of batting performance. By removing them from the dataset, we can eliminate unnecessary personal information and narrow the scope of the project to the pertinent variables.

In [399]:
# List of personal columns that are not relevant
personal_columns = ['nameGiven','birthState', 'birthCity', 'deathYear', 'deathMonth', 'deathDay', 'deathCountry', 'deathState', 'deathCity']

In [400]:
# Drop personal columns
raw_people.drop(personal_columns, axis=1, inplace=True)

Some pre cleaning

In [401]:
# Remove special characters and punctuation from nameFirst and nameLast
raw_people['nameFirst'] = raw_people['nameFirst'].str.replace(r"[\"\',]", '')
raw_people['nameLast'] = raw_people['nameLast'].str.replace(r"[\"\',]", '')

raw_people['nameFirst'] = raw_people['nameFirst'].str.replace('.', '')
raw_people['nameLast'] = raw_people['nameLast'].str.replace('.', '')

raw_people['nameFirst'] = raw_people['nameFirst'].str.replace(' ', '')

C:\Users\Hector G. Guerrero\AppData\Local\Temp\ipykernel_29804\3214174007.py:2: FutureWarning: The default value of regex will change from True to False in a future version.
  raw_people['nameFirst'] = raw_people['nameFirst'].str.replace(r"[\"\',]", '')
C:\Users\Hector G. Guerrero\AppData\Local\Temp\ipykernel_29804\3214174007.py:3: FutureWarning: The default value of regex will change from True to False in a future version.
  raw_people['nameLast'] = raw_people['nameLast'].str.replace(r"[\"\',]", '')
C:\Users\Hector G. Guerrero\AppData\Local\Temp\ipykernel_29804\3214174007.py:5: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  raw_people['nameFirst'] = raw_people['nameFirst'].str.replace('.', '')
C:\Users\Hector G. Guerrero\AppData\Local\Temp\ipykernel_29804\3214174007.py:6: FutureWarning: The default value of regex will change from 

In [402]:
# Shape
raw_people.shape

(20811, 15)

In [403]:
# Create a copy and save it to csv
people_pre = raw_people.copy()
people_pre.to_csv('pre_datasets/people_pre.csv', index=False)

### 2.4 Salary Dataset <a class="anchor" id="salary-dataset"></a>

In [404]:
# Shape and first rows
print(f'raw_salaries shape: {raw_salaries.shape}')
print(f'Rows: {raw_salaries.shape[0]}')
print(f'Columns: {raw_salaries.shape[1]}')
raw_salaries.head() 

raw_salaries shape: (46450, 15)
Rows: 46450
Columns: 15


firstname lastname  playerid   mlbid  year  salary              TeamName  \
0     David  Aardsma     25001  430911  2004  300000  San Francisco Giants   
1     David  Aardsma     25001  430911  2006  327000          Chicago Cubs   
2     David  Aardsma     25001  430911  2007  387500     Chicago White Sox   
3     David  Aardsma     25001  430911  2008  403250        Boston Red Sox   
4     David  Aardsma     25001  430911  2009  419000      Seattle Mariners   

   age  leaguerank  teamrank  averagesalary  leagueminimum serviceTime  \
0   22           0         0        2313535         300000        Null   
1   24           0         0        2699292         327000        Null   
2   25         675        23        2824751         380000        Null   
3   26         632        27        2925679         390000        Null   
4   27         635        20        2996106         400000        Null   

     borndate           name  
0  1981-12-27  David Aardsma  
1  1981-12-27  David Aardsma  
2  1981-12-27  David Aardsma  
3  1981-12-27  David Aardsma  
4  1981-12-27  David Aardsma

##### Dictionary

| Column Name | Description |
| --- | --- |
| firstname | The player's First name |
| lastname | The player's Last name |
| playerid | A unique identifier for each player |
| mlbid | A unique identifier for each player |
| year | Year of the salary |
| salary | Player salary for the year 
| age | The player's age during the year of the salary |
| leaguerank | The player's rank in the league based on salary |
| teamrank | The player's rank in the team based on salary |
| averagesalary | The average salary in the league for the year |
| leagueminimun | The minimum salary in the league for the year |
| serviceTime | The player's service time in the league |
| borndate | The player's birth date |

The salaries dataset provides detailed information about the salaries of baseball players. This dataset contains 46,450 rows and 14 columns. Each row represents a player's salary for a specific year, and the columns provide various details about the salary, the player, and the team they played for.

This dataset will be crucial for our analysis as it provides the target variable we want to predict - the player's salary. It also provides context about how the player's salary compares to others in the league and on their team.

Upon initial inspection of the salaries dataset, we've noticed that the `serviceTime` column contains several null values.

Additionally, this column doesn't seem to provide any crucial information for our analysis. Therefore, we'll drop this column from the dataset.

In [405]:
# Null values
raw_salaries.isnull().sum()

firstname        0
lastname         0
playerid         0
mlbid            0
year             0
salary           0
TeamName         0
age              0
leaguerank       0
teamrank         0
averagesalary    0
leagueminimum    0
serviceTime      0
borndate         0
name             0
dtype: int64

In [406]:
# 'Null' in serviceTime column
raw_salaries[raw_salaries['serviceTime'] == 'Null'].shape

(35121, 15)

It looks that there are no Null values in the dataset, 'Null' in serviceTime is just a string. There are 35,121 of them .We are going to drop the serviceTime column as it doesn't provide any crucial information for our analysis.

In [407]:
# Drop serviceTime column
raw_salaries.drop('serviceTime', axis=1, inplace=True)

In [408]:
# Data types
raw_salaries.dtypes

firstname        object
lastname         object
playerid          int64
mlbid             int64
year              int64
salary            int64
TeamName         object
age               int64
leaguerank        int64
teamrank          int64
averagesalary     int64
leagueminimum     int64
borndate         object
name             object
dtype: object

Some pre cleaning

In [409]:
# Remove special characters and punctuation from firstName and lastName
raw_salaries['firstname'] = raw_salaries['firstname'].str.replace(r"[\"\',]", '')
raw_salaries['lastname'] = raw_salaries['lastname'].str.replace(r"[\"\',]", '')

raw_salaries['firstname'] = raw_salaries['firstname'].str.replace('.', '')
raw_salaries['lastname'] = raw_salaries['lastname'].str.replace('.', '')


C:\Users\Hector G. Guerrero\AppData\Local\Temp\ipykernel_29804\3092410976.py:2: FutureWarning: The default value of regex will change from True to False in a future version.
  raw_salaries['firstname'] = raw_salaries['firstname'].str.replace(r"[\"\',]", '')
C:\Users\Hector G. Guerrero\AppData\Local\Temp\ipykernel_29804\3092410976.py:3: FutureWarning: The default value of regex will change from True to False in a future version.
  raw_salaries['lastname'] = raw_salaries['lastname'].str.replace(r"[\"\',]", '')
C:\Users\Hector G. Guerrero\AppData\Local\Temp\ipykernel_29804\3092410976.py:5: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  raw_salaries['firstname'] = raw_salaries['firstname'].str.replace('.', '')
C:\Users\Hector G. Guerrero\AppData\Local\Temp\ipykernel_29804\3092410976.py:6: FutureWarning: The default value of regex will 

##### Full name column
In our current dataset, player names are split into two separate columns: `firstname` and `lastname`. While this format can be useful for certain types of analysis, it might be more convenient for us to have a single column that contains the full name of each player.

In [410]:
# concat first and last name
raw_salaries['name'] = raw_salaries['firstname'] + ' ' + raw_salaries['lastname']

raw_salaries.head()

firstname lastname  playerid   mlbid  year  salary              TeamName  \
0     David  Aardsma     25001  430911  2004  300000  San Francisco Giants   
1     David  Aardsma     25001  430911  2006  327000          Chicago Cubs   
2     David  Aardsma     25001  430911  2007  387500     Chicago White Sox   
3     David  Aardsma     25001  430911  2008  403250        Boston Red Sox   
4     David  Aardsma     25001  430911  2009  419000      Seattle Mariners   

   age  leaguerank  teamrank  averagesalary  leagueminimum    borndate  \
0   22           0         0        2313535         300000  1981-12-27   
1   24           0         0        2699292         327000  1981-12-27   
2   25         675        23        2824751         380000  1981-12-27   
3   26         632        27        2925679         390000  1981-12-27   
4   27         635        20        2996106         400000  1981-12-27   

            name  
0  David Aardsma  
1  David Aardsma  
2  David Aardsma  
3  David Aardsma  
4  David Aardsma

##### `borndate` column
The `borndate` column contains the birth date of each player. Currently this column is in string format, which is not ideal for our analysis. We'll convert this column to a datetime format.

In [411]:
# borndate to datetime
raw_salaries['borndate'] = pd.to_datetime(raw_salaries['borndate'])

In [412]:
# Drop rows with year = 2023
raw_salaries = raw_salaries[raw_salaries['year'] != 2023]

In [413]:
drop_rows = ['ChJi197', 'ChTr196']

In [414]:
# Create a copy and save it to csv
salaries_pre = raw_salaries.copy()
salaries_pre.to_csv('pre_datasets/salaries_pre.csv', index=False)

### 2.5 Teams Dataset <a class="anchor" id="teams-dataset"></a>

In [415]:
# Shape and first rows
print(f'raw_teams shape: {raw_teams.shape}')
print(f'Rows: {raw_teams.shape[0]}')
print(f'Columns: {raw_teams.shape[1]}')
raw_teams.sample(5)

raw_teams shape: (3015, 47)
Rows: 3015
Columns: 47


yearID lgID teamID franchID divID  Rank    G  Ghome   W   L DivWin  \
2533    2006   AL    TOR      TOR     E     2  162   81.0  87  75      N   
866     1930   AL    CHA      CHW   NaN     7  154   78.0  62  92    NaN   
2151    1993   AL    TEX      TEX     W     2  162   81.0  86  76      N   
1271    1955   AL    KC1      OAK   NaN     6  155   76.0  63  91    NaN   
2516    2006   NL    HOU      HOU     C     2  162   81.0  82  80      N   

     LgWin WSWin    R    AB     H   2B  3B   HR   BB      SO     SB    CS  \
2533     N     N  809  5596  1591  348  27  199  514   906.0   65.0  33.0   
866      N     N  729  5419  1496  256  90   63  389   479.0   74.0  40.0   
2151     N     N  835  5510  1472  284  39  181  483   984.0  113.0  67.0   
1271     N     N  638  5335  1395  189  46  121  463   725.0   22.0  36.0   
2516     N     N  735  5521  1407  275  27  174  585  1076.0   79.0  36.0   

       HBP    SF   RA   ER   ERA  CG  SHO  SV  IPouts    HA  HRA  BBA   SOA  \
2533  63.0  52.0  754  694  4.37   6    6  42    4285  1447  185  504  1076   
866    NaN   NaN  884  712  4.71  63    2  10    4083  1629   74  407   471   
2151  48.0  56.0  751  684  4.28  20    6  45    4315  1476  144  562   957   
1271   NaN   NaN  911  822  5.35  29    9  22    4146  1486  175  707   572   
2516  73.0  46.0  719  666  4.08   5   12  42    4406  1425  182  480  1160   

        E   DP     FP                   name                 park  attendance  \
2533   99  157  0.984      Toronto Blue Jays        Rogers Centre   2302212.0   
866   233  136  0.962      Chicago White Sox        Comiskey Park    406123.0   
2151  132  145  0.979          Texas Rangers    Arlington Stadium   2244616.0   
1271  146  174  0.976  Kansas City Athletics  Municipal Stadium I   1393054.0   
2516   80  164  0.987         Houston Astros     Minute Maid Park   3022763.0   

      BPF  PPF teamIDBR teamIDlahman45 teamIDretro  
2533  100  100      TOR            TOR         TOR  
866    96   99      CHW            CHA         CHA  
2151   96   96      TEX            TEX         TEX  
1271  100  105      KCA            KC1         KC1  
2516  100   99      HOU            HOU         HOU

##### Dictionary

| Column Name | Description |
| --- | --- |
| yearID | Year (Season played) |
| lgID | The league the player played in |
| teamID | ID of the team they played for the season |
| franchID | Franchise ID |
| divID | Division ID |
| Rank | Team rank at the end of the season |
| G | Games played |
| Ghome | Games played at home |
| W | Wins |
| L | Losses |
| DivWin | Division Winner (Y or N) |
| WCWin | Wild Card Winner (Y or N) |
| LgWin | League Champion(Y or N) |
| WSWin | World Series Winner (Y or N) |
| R | Total number of runs scored by the team in the season |
| AB | Total number of at bats by the team in the season |
| H | Total number of hits by the team in the season |
| 2B | Total number of doubles by the team in the season |
| 3B | Total number of triples by the team in the season |
| HR | Total number of home runs by the team in the season |
| BB | Total number of walks by the team in the season |
| SO | Total number of strikeouts by the team in the season |
| SB | Total number of stolen bases by the team in the season |
| CS | Total number of times caught stealing by the team in the season |
| HBP | Total number of times hit by pitch by the team in the season |
| SF | Total number of sacrifice flies by the team in the season |
| RA | Total number of runs allowed by the team in the season |
| ER | Total number of earned runs allowed by the team in the season |
| ERA | Earned run average |
| CG | Total number of complete games pitched by the team in the season |
| SHO | Total number of shutouts pitched by the team in the season |
| SV | Total number of saves by the team in the season |
| IPouts | Total number of outs pitched by the team in the season |
| HA | Total number of hits allowed by the team in the season |
| HRA | Total number of home runs allowed by the team in the season |
| BBA | Total number of walks allowed by the team in the season |
| SOA | Total number of strikeouts by the team in the season |
| E | Total number of errors by the team in the season |
| DP | Total number of double plays turned by the team in the season |
| FP | Fielding percentage |
| name | Team name |
| park | Team park |
| attendance | Total attendance for the season |
| BPF | Three-year park factor for batters |
| PPF | Three-year park factor for pitchers |
| teamIDBR | Team ID used by Baseball Reference website |
| teamIDlahman45 | Team ID used in Lahman database version 4.5 |
| teamIDretro | Team ID used by Retrosheet |

The `raw_teams` dataset is a comprehensive collection of team-based statistics for each season. It has 3015 rows and 47 columns, each row representing a team's performance in a particular season.

For the time being, we'll keep all the columns as they might provide useful insights for our analysis. We can always drop unnecessary columns later once we understand the data better.

In [416]:
# Data types
raw_teams.dtypes

yearID              int64
lgID               object
teamID             object
franchID           object
divID              object
Rank                int64
G                   int64
Ghome             float64
W                   int64
L                   int64
DivWin             object
LgWin              object
WSWin              object
R                   int64
AB                  int64
H                   int64
2B                  int64
3B                  int64
HR                  int64
BB                  int64
SO                float64
SB                float64
CS                float64
HBP               float64
SF                float64
RA                  int64
ER                  int64
ERA               float64
CG                  int64
SHO                 int64
SV                  int64
IPouts              int64
HA                  int64
HRA                 int64
BBA                 int64
SOA                 int64
E                   int64
DP                  int64
FP          

It looks that all the data types are correct, so we can move on to the next step.

In [417]:
# Create a copy and save it to csv
teams_pre = raw_teams.copy()
teams_pre.to_csv('pre_datasets/teams_pre.csv', index=False)

## 3. Table Joins <a class="anchor" id="table-joins"></a>

Now, let's join our dataframes to create a single dataset that contains player-level information for each season. The resulting dataset will be used for further analysis and modeling.

The goal here is to find a common link between the datasets that can be used to join them together. This might require some trial and error, and possibly some data cleaning to ensure the keys match up correctly. This process can turn iterative.

In [418]:
pd.set_option('display.max_columns', None)

def show_player_info(dataset, player_name):
    '''Display player information for sanity check'''

    player_info = dataset[dataset['name_common'] == player_name]
    
    return player_info

##### WAR and People datasets

In [419]:
print(war_pre.shape)
print(people_pre.shape)

(103491, 16)
(20811, 15)


In [420]:
# Join people_pre with war_pre on bbrefID and player_ID
war_people = war_pre.merge(people_pre, how='left',left_on='player_ID', right_on='bbrefID')
war_people.sample(5)

name_common   age  year_ID  player_ID pitcher    mlb_ID  stint_ID  \
76187    Neal Ball  30.0     1911   ballne01       N  110513.0         1   
78470    Pat Lyons  30.0     1890  lyonspa01       N  118052.0         1   
56264   Joe Tinker  34.0     1915  tinkejo01       N  123353.0         1   
19458  Chris Young  26.0     2005  youngch03       Y  432934.0         1   
36826  Fred Dunlap  32.0     1891  dunlafr01       N  113630.0         1   

          PA    G    Inn    salary    OPS_plus   WAR  WAR_def  WAR_off  \
76187  455.0  116    0.0       0.0  103.889318  1.68     0.21     1.50   
78470   42.0   11    0.0       0.0  -33.602463 -0.62    -0.18    -0.43   
56264   82.0   31    0.0   12000.0  115.711779  0.59     0.26     0.46   
19458    5.0    2  164.7  400000.0 -100.000000 -0.08     0.00    -0.08   
36826   31.0    8    0.0       0.0   97.903305  0.03    -0.06     0.11   

      team_ID   playerID  birthYear  birthMonth  birthDay birthCountry  \
76187     CLE   ballne01     1881.0         4.0      22.0          USA   
78470     CLV  lyonspa01     1860.0         3.0       NaN          CAN   
56264     CHI  tinkejo01     1880.0         7.0      27.0          USA   
19458     TEX  youngch03     1979.0         5.0      25.0          USA   
36826     WAS  dunlafr01     1859.0         5.0      21.0          USA   

      nameFirst nameLast  weight  height bats throws       debut   finalGame  \
76187      Neal     Ball   145.0    67.0    R      R  1907-09-11  1913-06-30   
78470       Pat    Lyons   155.0    67.0    R      R  1890-07-21  1890-08-05   
56264       Joe   Tinker   175.0    69.0    R      R  1902-04-17  1916-09-22   
19458     Chris    Young   255.0    82.0    R      R  2004-08-24  2017-06-17   
36826      Fred   Dunlap   165.0    68.0    R      R  1880-05-01  1891-04-20   

        retroID    bbrefID  
76187  balln101   ballne01  
78470  lyonp101  lyonspa01  
56264  tinkj101  tinkejo01  
19458  younc003  youngch03  
36826  dunlf101  dunlafr01

In [421]:
show_player_info(war_people, 'Joey Gallo')

name_common   age  year_ID  player_ID pitcher     mlb_ID  stint_ID  \
56527  Joey Gallo  21.0     2015  gallojo01       N   608336.0         1   
56528  Joey Gallo  22.0     2016  gallojo01       N   608336.0         1   
56529  Joey Gallo  23.0     2017  gallojo01       N   608336.0         1   
56530  Joey Gallo  24.0     2018  gallojo01       N   608336.0         1   
56531  Joey Gallo  25.0     2019  gallojo01       N   608336.0         1   
56532  Joey Gallo  26.0     2020  gallojo01       N   608336.0         1   
56533  Joey Gallo  27.0     2021  gallojo01       N  1216672.0         2   
56534  Joey Gallo  28.0     2022  gallojo01       N  1216672.0         2   

          PA    G     Inn      salary    OPS_plus   WAR  WAR_def  WAR_off  \
56527  123.0   36   264.0         0.0   91.332137  0.32     0.07     0.21   
56528   30.0   17    72.0         0.0   -2.863830 -0.52    -0.24    -0.32   
56529  532.0  145  1208.0    537120.0  118.426407  2.92    -0.63     3.32   
56530  577.0  148  1199.0    560000.0  108.778664  2.41    -0.01     1.86   
56531  297.0   70   614.3    605500.0  144.925708  3.08     0.47     2.54   
56532  226.0   57   475.7   4400000.0   86.573416  1.51     1.11     0.20   
56533  616.0  153  1304.4   6200000.0  230.520541  4.61     0.92     3.04   
56534  410.0  126   944.4  10275000.0  162.570976  0.23    -0.13    -0.18   

      team_ID   playerID  birthYear  birthMonth  birthDay birthCountry  \
56527     TEX  gallojo01     1993.0        11.0      19.0          USA   
56528     TEX  gallojo01     1993.0        11.0      19.0          USA   
56529     TEX  gallojo01     1993.0        11.0      19.0          USA   
56530     TEX  gallojo01     1993.0        11.0      19.0          USA   
56531     TEX  gallojo01     1993.0        11.0      19.0          USA   
56532     TEX  gallojo01     1993.0        11.0      19.0          USA   
56533     TEX  gallojo01     1993.0        11.0      19.0          USA   
56534     NYY  gallojo01     1993.0        11.0      19.0          USA   

      nameFirst nameLast  weight  height bats throws       debut   finalGame  \
56527      Joey    Gallo   250.0    77.0    L      R  2015-06-02  2023-06-24   
56528      Joey    Gallo   250.0    77.0    L      R  2015-06-02  2023-06-24   
56529      Joey    Gallo   250.0    77.0    L      R  2015-06-02  2023-06-24   
56530      Joey    Gallo   250.0    77.0    L      R  2015-06-02  2023-06-24   
56531      Joey    Gallo   250.0    77.0    L      R  2015-06-02  2023-06-24   
56532      Joey    Gallo   250.0    77.0    L      R  2015-06-02  2023-06-24   
56533      Joey    Gallo   250.0    77.0    L      R  2015-06-02  2023-06-24   
56534      Joey    Gallo   250.0    77.0    L      R  2015-06-02  2023-06-24   

        retroID    bbrefID  
56527  gallj002  gallojo01  
56528  gallj002  gallojo01  
56529  gallj002  gallojo01  
56530  gallj002  gallojo01  
56531  gallj002  gallojo01  
56532  gallj002  gallojo01  
56533  gallj002  gallojo01  
56534  gallj002  gallojo01

##### WAR/People and Batting datasets

In [422]:
# shape
war_people.shape

(103491, 31)

In [423]:
batting_pre.shape

(103693, 20)

In [424]:
# Join war_people and batting_pre on player_ID = playerID , year_ID = yearID
war_batting_people = war_people.merge(batting_pre, how='left', left_on=['playerID', 'year_ID'], right_on=['playerID', 'yearID'])
war_batting_people.sample(5)

name_common   age  year_ID  player_ID pitcher    mlb_ID  stint_ID  \
94481     Tex Hughson  31.0     1947  hughste01       Y  116295.0         1   
72481  Mickey Stanley  25.0     1968  stanlmi01       N  122672.0         1   
1039        Al Bumbry  27.0     1974  bumbral01       N  111687.0         1   
63134    Ken McMullen  29.0     1971  mcmulke01       N  118834.0         1   
50261   Jeff Robinson  29.0     1991  robinje02       Y  121316.0         1   

          PA  G_x     Inn    salary    OPS_plus   WAR  WAR_def  WAR_off  \
94481   70.0   29     0.0   25000.0  -57.206321 -0.59    -0.01    -0.58   
72481  641.0  153  1293.3   15500.0  101.988944  3.74     0.99     2.73   
1039   298.0   94   617.0       0.0   73.218430 -0.09    -0.26    -0.21   
63134  657.0  160  1428.0   40000.0  107.028506  4.24     1.09     3.58   
50261    0.0    0   104.3  575000.0    0.000000  0.00     0.00     0.00   

      team_ID   playerID  birthYear  birthMonth  birthDay birthCountry  \
94481     BOS  hughste01     1916.0         2.0       9.0          USA   
72481     DET  stanlmi01     1942.0         7.0      20.0          USA   
1039      BAL  bumbral01     1947.0         4.0      21.0          USA   
63134     CAL  mcmulke01     1942.0         6.0       1.0          USA   
50261     BAL  robinje02     1961.0        12.0      14.0          USA   

      nameFirst  nameLast  weight  height bats throws       debut   finalGame  \
94481       Tex   Hughson   198.0    75.0    R      R  1941-04-16  1949-10-02   
72481    Mickey   Stanley   185.0    73.0    R      R  1964-09-13  1978-09-28   
1039         Al    Bumbry   170.0    68.0    L      R  1972-09-05  1985-10-05   
63134       Ken  McMullen   190.0    75.0    R      R  1962-09-17  1977-09-14   
50261      Jeff  Robinson   210.0    78.0    R      R  1987-04-12  1992-07-20   

        retroID    bbrefID  yearID    G_y     AB     R      H    2B   3B  \
94481  hught102  hughste01  1947.0   29.0   61.0   3.0    2.0   1.0  0.0   
72481  stanm101  stanlmi01  1968.0  153.0  583.0  88.0  151.0  16.0  6.0   
1039   bumba001  bumbral01  1974.0   94.0  270.0  35.0   63.0  10.0  3.0   
63134  mcmuk101  mcmulke01  1971.0  160.0  593.0  63.0  148.0  19.0  2.0   
50261  robij002  robinje02  1991.0   21.0    0.0   0.0    0.0   0.0  0.0   

         HR   RBI    SB   CS    BB    SO   IBB  HBP   SH   SF  GIDP teamID  
94481   0.0   2.0   0.0  0.0   5.0  22.0   0.0  0.0  4.0  0.0   1.0    BOS  
72481  11.0  60.0   4.0  3.0  42.0  57.0   1.0  4.0  8.0  4.0  22.0    DET  
1039    1.0  19.0  12.0  4.0  21.0  46.0   2.0  1.0  3.0  3.0   3.0    BAL  
63134  21.0  68.0   1.0  1.0  53.0  74.0  10.0  3.0  4.0  4.0  18.0    CAL  
50261   0.0   0.0   0.0  0.0   0.0   0.0   0.0  0.0  0.0  0.0   0.0    BAL

In [425]:
show_player_info(war_batting_people, 'Joey Gallo')

name_common   age  year_ID  player_ID pitcher     mlb_ID  stint_ID  \
56527  Joey Gallo  21.0     2015  gallojo01       N   608336.0         1   
56528  Joey Gallo  22.0     2016  gallojo01       N   608336.0         1   
56529  Joey Gallo  23.0     2017  gallojo01       N   608336.0         1   
56530  Joey Gallo  24.0     2018  gallojo01       N   608336.0         1   
56531  Joey Gallo  25.0     2019  gallojo01       N   608336.0         1   
56532  Joey Gallo  26.0     2020  gallojo01       N   608336.0         1   
56533  Joey Gallo  27.0     2021  gallojo01       N  1216672.0         2   
56534  Joey Gallo  28.0     2022  gallojo01       N  1216672.0         2   

          PA  G_x     Inn      salary    OPS_plus   WAR  WAR_def  WAR_off  \
56527  123.0   36   264.0         0.0   91.332137  0.32     0.07     0.21   
56528   30.0   17    72.0         0.0   -2.863830 -0.52    -0.24    -0.32   
56529  532.0  145  1208.0    537120.0  118.426407  2.92    -0.63     3.32   
56530  577.0  148  1199.0    560000.0  108.778664  2.41    -0.01     1.86   
56531  297.0   70   614.3    605500.0  144.925708  3.08     0.47     2.54   
56532  226.0   57   475.7   4400000.0   86.573416  1.51     1.11     0.20   
56533  616.0  153  1304.4   6200000.0  230.520541  4.61     0.92     3.04   
56534  410.0  126   944.4  10275000.0  162.570976  0.23    -0.13    -0.18   

      team_ID   playerID  birthYear  birthMonth  birthDay birthCountry  \
56527     TEX  gallojo01     1993.0        11.0      19.0          USA   
56528     TEX  gallojo01     1993.0        11.0      19.0          USA   
56529     TEX  gallojo01     1993.0        11.0      19.0          USA   
56530     TEX  gallojo01     1993.0        11.0      19.0          USA   
56531     TEX  gallojo01     1993.0        11.0      19.0          USA   
56532     TEX  gallojo01     1993.0        11.0      19.0          USA   
56533     TEX  gallojo01     1993.0        11.0      19.0          USA   
56534     NYY  gallojo01     1993.0        11.0      19.0          USA   

      nameFirst nameLast  weight  height bats throws       debut   finalGame  \
56527      Joey    Gallo   250.0    77.0    L      R  2015-06-02  2023-06-24   
56528      Joey    Gallo   250.0    77.0    L      R  2015-06-02  2023-06-24   
56529      Joey    Gallo   250.0    77.0    L      R  2015-06-02  2023-06-24   
56530      Joey    Gallo   250.0    77.0    L      R  2015-06-02  2023-06-24   
56531      Joey    Gallo   250.0    77.0    L      R  2015-06-02  2023-06-24   
56532      Joey    Gallo   250.0    77.0    L      R  2015-06-02  2023-06-24   
56533      Joey    Gallo   250.0    77.0    L      R  2015-06-02  2023-06-24   
56534      Joey    Gallo   250.0    77.0    L      R  2015-06-02  2023-06-24   

        retroID    bbrefID  yearID    G_y     AB     R      H    2B   3B  \
56527  gallj002  gallojo01  2015.0   36.0  108.0  16.0   22.0   3.0  1.0   
56528  gallj002  gallojo01  2016.0   17.0   25.0   2.0    1.0   0.0  0.0   
56529  gallj002  gallojo01  2017.0  145.0  449.0  85.0   94.0  18.0  3.0   
56530  gallj002  gallojo01  2018.0  148.0  500.0  82.0  103.0  24.0  1.0   
56531  gallj002  gallojo01  2019.0   70.0  241.0  54.0   61.0  15.0  1.0   
56532  gallj002  gallojo01  2020.0   57.0  193.0  23.0   35.0   8.0  0.0   
56533  gallj002  gallojo01  2021.0  153.0  498.0  90.0   99.0  13.0  1.0   
56534  gallj002  gallojo01  2022.0  126.0  350.0  48.0   56.0   8.0  2.0   

         HR   RBI   SB   CS     BB     SO  IBB  HBP   SH   SF  GIDP teamID  
56527   6.0  14.0  3.0  0.0   15.0   57.0  3.0  0.0  0.0  0.0   0.0    TEX  
56528   1.0   1.0  1.0  0.0    5.0   19.0  0.0  0.0  0.0  0.0   0.0    TEX  
56529  41.0  80.0  7.0  2.0   75.0  196.0  1.0  8.0  0.0  0.0   3.0    TEX  
56530  40.0  92.0  3.0  4.0   74.0  207.0  4.0  3.0  0.0  0.0   3.0    TEX  
56531  22.0  49.0  4.0  2.0   52.0  114.0  4.0  2.0  1.0  1.0   0.0    TEX  
56532  10.0  26.0  2.0  0.0   29.0   79.0  2.0  4.0  0.0  0.0   0.0    TEX  
56533  38.0  77

In [426]:
war_batting_people.shape

(103491, 50)

##### WAR/People/Batting and Team datasets

There are several columns for team identifiers, let's decide on which one to use.

In [427]:
# Compare the teamIDBR column from teams_pre to the team_ID column from war_batting_people
print(f'teamIDBR (teams_pre) vs team_ID (war_batting_people):\n {teams_pre["teamIDBR"].isin(war_batting_people["team_ID"]).value_counts()}\n')

# Compare the teamID column from teams_pre to the teamID column from war_batting_people 
print(f'teamID (teams_pre) vs teamID (war_batting_people):\n {teams_pre["teamID"].isin(war_batting_people["teamID"]).value_counts()}\n')

# Compare the teamID column from temas_pre to the team_ID column from war_batting_people
print(f'teamID (teams_pre) vs team_ID (war_batting_people):\n {teams_pre["teamID"].isin(war_batting_people["team_ID"]).value_counts()}')

teamIDBR (teams_pre) vs team_ID (war_batting_people):
 True    3015
Name: teamIDBR, dtype: int64

teamID (teams_pre) vs teamID (war_batting_people):
 True    3015
Name: teamID, dtype: int64

teamID (teams_pre) vs team_ID (war_batting_people):
 True     1735
False    1280
Name: teamID, dtype: int64


In [428]:
# Unique values in war_batting_people team_ID
war_batting_people['teamID'].nunique()

149

In [429]:
# Unique values in teams_pre teamID 
teams_pre['teamID'].nunique()

149

In [430]:
war_batting_people.shape

(103491, 50)

In [431]:
# Join war_batting_people and raw_teams on teamID = teamID and yearID = yearID
war_batting_people_teams = war_batting_people.merge(teams_pre, how='left', left_on=['teamID', 'yearID'], right_on=['teamID', 'yearID'])

In [432]:
war_batting_people_teams.sample(5)

name_common   age  year_ID  player_ID pitcher    mlb_ID  stint_ID  \
15443    Camilo Pascual  37.0     1971  pascuca02       Y  120266.0         1   
27084     Dexter Fowler  33.0     2019  fowlede01       N  451594.0         1   
88114     Russ Van Atta  31.0     1937  vanatru01       Y  123625.0         1   
103466  Zoilo Versalles  25.0     1965  versazo01       N  123713.0         1   
10505      Bob OFarrell  21.0     1918  o'farbo01       N  119947.0         1   

           PA  G_x     Inn      salary    OPS_plus   WAR  WAR_def  WAR_off  \
15443     6.0    9    23.3     28000.0  280.659036  0.22     0.00     0.22   
27084   574.0  150  1138.0  16500000.0   99.644554  1.42    -0.80     1.89   
88114    15.0   16     0.0      6000.0  206.462672  0.36     0.00     0.36   
103466  728.0  160  1406.3     28000.0  115.213464  7.17     3.01     5.33   
10505   125.0   52     0.0         0.0  132.718278  0.80    -0.04     1.06   

       team_ID   playerID  birthYear  birthMonth  birthDay birthCountry  \
15443      CLE  pascuca02     1934.0         1.0      20.0         Cuba   
27084      STL  fowlede01     1986.0         3.0      22.0          USA   
88114      SLB  vanatru01     1906.0         6.0      21.0          USA   
103466     MIN  versazo01     1939.0        12.0      18.0         Cuba   
10505      CHC  ofarrbo01     1896.0        10.0      19.0          USA   

       nameFirst   nameLast  weight  height bats throws       debut  \
15443     Camilo    Pascual   170.0    71.0    R      R  1954-04-15   
27084     Dexter     Fowler   205.0    77.0    B      R  2008-09-02   
88114       Russ   Van Atta   184.0    72.0    L      L  1933-04-25   
103466     Zoilo  Versalles   146.0    70.0    R      R  1959-08-01   
10505        Bob   OFarrell   180.0    69.0    R      R  1915-09-05   

         finalGame   retroID    bbrefID  yearID    G_y   AB_x    R_x    H_x  \
15443   1971-05-05  pascc102  pascuca02  1971.0    9.0    5.0    0.0    3.0   
27084   2021-04-09  fowld001  fowlede01  2019.0  150.0  487.0   69.0  116.0   
88114   1939-05-11  vanar101  vanatru01  1937.0   16.0   13.0    3.0    6.0   
103466  1971-09-28  versz101  versazo01  1965.0  160.0  666.0  126.0  182.0   
10505   1935-09-23  ofarb101  o'farbo01  1918.0   52.0  113.0    9.0   32.0   

        2B_x  3B_x  HR_x   RBI  SB_x  CS_x  BB_x   SO_x  IBB  HBP_x   SH  \
15443    1.0   0.0   0.0   0.0   0.0   0.0   0.0    1.0  0.0    0.0  1.0   
27084   24.0   1.0  19.0  67.0   8.0   5.0  74.0  142.0  1.0    8.0  1.0   
88114    0.0   0.0   1.0   1.0   0.0   0.0   2.0    2.0  0.0    0.0  0.0   
103466  45.0  12.0  19.0  77.0  27.0   5.0  41.0  122.0  3.0    7.0  6.0   
10505    7.0   3.0   1.0  14.0   0.0   0.0  10.0   15.0  0.0    1.0  1.0   

        SF_x  GIDP teamID lgID franchID divID  Rank      G  Ghome      W  \
15443    0.0   0.0    CLE   AL      CLE     E   6.0  162.0   81.0   60.0   
27084    4.0   6.0    SLN   NL      STL     C   1.0  162.0   81.0   91.0   
88114    0.0   0.0    SLA   AL      BAL   NaN   8.0  156.0   78.0   46.0   
103466   8.0   7.0    MIN   AL      MIN   NaN   1.0  162.0   81.0  102.0   
10505    0.0   0.0    CHN   NL      CHC   NaN   1.0  131.0   74.0   84.0   

            L DivWin LgWin WSWin    R_y    AB_y     H_y   2B_y  3B_y   HR_y  \
15443   102.0      N     N     N  543.0  5467.0  1303.0  200.0  20.0  109.0   
27084    71.0      Y     N     N  764.0  5449.0  1336.0  246.0  24.0  210.0   
88114   108.0    NaN     N     N  715.0  5510.0  1573.0  327.0  44.0   71.0   
103466   60.0    NaN     Y     N  774.0  5488.0  1396.0  257.0  42.0  150.0   
10505    45.0    NaN     Y     N  538.0  4325.0  1147.0  164.0  53.0   21.0   

         BB_y    SO_y   SB_y  CS_y  HBP_y  SF_y      RA     ER   ERA    CG  \
15443   467.0   868.0   57.0  37.0   31.0  29.0   747.0  684.0  4.28  21.0   
27084   561.0  1420.0  116.0  29.0   76.0  39.0   662.0  609.0  3.80   1.0   
88114   514.0   510.0   30.0  27.0    NaN   NaN  1023.0  909.0  6.00  55.0   
10

In [433]:
war_batting_people_teams.shape

(103491, 95)

__Some Pre cleaning__

Rows that we realized during this step that contain null values and dropping them would not affect the analysis, as they are not relevant players with relevant stats.

In [434]:
# Pre cleaning - Rows that we realized during this step that contain null values and dropping them would not affect the analysis, as they are not relevant players with relevant stats.
rows_to_drop = ['sternad01', 
                'hegmabo01', 
                'jimerch01', 
                'tremich01',
                'firovda01',
                'singldu01',
                'lunarfe01',
                'matosfr01',
                'reyesgi01',
                'manrifr01',
                'krugeja01',
                'pankoji01',
                'hietpjo01',
                'morgake01',
                'dalesma01',
                'mooreke02',
                'davidma01',
                'davisma02',
                'calzana01',
                'santape01',
                'carabra01',
                'mckeewa01',
                'esposbr01',
                ]

In [435]:
# Drop rows
war_batting_people_teams = war_batting_people_teams[~war_batting_people_teams['playerID'].isin(rows_to_drop)]

In [436]:
special_drop_dict = {'stewach01': [2007, 2008, 2010],
                     'freesda01': [2009],
                     'disarga01': [1989],
                     'kingsge01': [1996],
                     'snowjt01': [2008],
                     'larueja01': [2009, 2010],
                     'posadjo01': [1995],
                     'thornlo01': [1985, 1990],
                     'zuvelpa01': [1991],
                     'murphse01': [2019],
                     'gorete01': [2020],
                     'lintz01': [2021],
                     'wilsova01': [1999],
                     }

In [438]:
# Drop rows in special_drop_dict
for key, value in special_drop_dict.items():
    war_batting_people_teams = war_batting_people_teams[~((war_batting_people_teams['playerID'] == key) & (war_batting_people_teams['yearID'].isin(value)))]

##### WAR/People/Batting/Team and Salary datasets

In order to align our datasets for comparison, we'll first create a subset of data starting from 1985, as this is the earliest year available in the `salaries_pre` dataset. This will allow us to identify common links between the datasets and determine the most suitable one for joining. Once we've made this decision, we'll proceed to join the full datasets. Any null values resulting from years prior to 1985 will be addressed in the next stages of the project.

In [439]:
# war_batting_people_teams starting in 1985
war_batting_people_teams_85 = war_batting_people_teams.copy()
war_batting_people_teams_85 = war_batting_people_teams[war_batting_people_teams['year_ID'] >= 1985]

`mlb_ID` and `mlbid` seem to be the only common link between the datasets. Let's see if we can use these columns to join the datasets.

In [440]:
# Compare mlb_ID from war_batting_people_teams to mlbid from salaries_pre
print(f'mlb_ID (war_batting_people_teams_85) vs mlbid (salaries_pre):\n {war_batting_people_teams_85["mlb_ID"].isin(salaries_pre["mlbid"]).value_counts()}')

mlb_ID (war_batting_people_teams_85) vs mlbid (salaries_pre):
 True     41501
False     4106
Name: mlb_ID, dtype: int64


It appears that there are 41,501 instances where the `mlb_ID` from the war_batting_people_teams_85 dataset matches the `mlbid` from the salaries_pre dataset. However, there are also 4,106 instances where the IDs do not match.

There could be inconsistencies in the way the IDs are recorded in the two datasets. Let's explore this further by looking at the IDs that don't match.

In [441]:
# Extract a sample player with different mlb_ID
difference_sample = war_batting_people_teams_85[~war_batting_people_teams_85["mlb_ID"].isin(salaries_pre["mlbid"])][['name_common', 'mlb_ID']].sample(1)
name_d = difference_sample.values[0][0]
mlbid_d = difference_sample.values[0][1]

print('From war_batting_people_teams_85:')
print(f'Name: {name_d}\nmlb_ID: {mlbid_d}\n')

print('From salaries_pre:')
print(salaries_pre[salaries_pre["name"] == name_d][["name", "mlbid"]].head(1).iloc[0])

From war_batting_people_teams_85:
Name: Dave Gallagher
mlb_ID: 229066.0

From salaries_pre:
name     Dave Gallagher
mlbid            114533
Name: 14091, dtype: object


It appears that there is a discrepancy between the `mlb_ID` in the war_batting_people_teams_85 dataset and the `mlbid` in the salaries_pre dataset. This could be due to a variety of reasons, such as data entry errors, different data sources, or changes in player IDs over time.

The best solution would be to find/create a common link between the datasets. We don't want to join by player's name as this is not a unique identifier, some players can share the same name. 

Let's explore this solution further.



##### - Creating a unique player ID for both datasets

Both datasets share birth dates information and names, we can play around with these columns to create a unique player ID for both datasets. This will be a trial and error and iterative process.

__Unique ID for war_batting_people_teams__

In [442]:
# Some of this columns I had dropped them before, had to go back and add them again

# Birth dates to int
war_batting_people_teams_85['birthYear'] = war_batting_people_teams_85['birthYear'].astype('Int64')
war_batting_people_teams_85['birthMonth'] = war_batting_people_teams_85['birthMonth'].astype('Int64')
war_batting_people_teams_85['birthDay'] = war_batting_people_teams_85['birthDay'].astype('Int64')

# Birth dates to int full dataset
war_batting_people_teams['birthYear'] = war_batting_people_teams['birthYear'].astype('Int64')
war_batting_people_teams['birthMonth'] = war_batting_people_teams['birthMonth'].astype('Int64')
war_batting_people_teams['birthDay'] = war_batting_people_teams['birthDay'].astype('Int64')
war_batting_people_teams['birthYear_3'] = war_batting_people_teams['birthYear'].astype(str).str[:3]

# First and second letter of nameFirst and nameLast 
war_batting_people_teams_85['nameFirst_2'] = war_batting_people_teams_85['nameFirst'].str[:2]
war_batting_people_teams_85['nameLast_2'] = war_batting_people_teams_85['nameLast'].str[:2]

# First and second letter of nameFirst and nameLast full dataset
war_batting_people_teams['nameFirst_2'] = war_batting_people_teams['nameFirst'].str[:2]
war_batting_people_teams['nameLast_2'] = war_batting_people_teams['nameLast'].str[:2]

# concat initials year, month and day for new ID
war_batting_people_teams_85['new_id_x'] = war_batting_people_teams_85['nameFirst_2'] + war_batting_people_teams_85['nameLast_2'] + war_batting_people_teams_85['birthYear'].astype(str) #+ war_batting_people_teams_85['birthMonth'].astype(str) + war_batting_people_teams_85['birthDay'].astype(str) # + war_batting_people_teams_85['birthYear'].astype(str)

# concat initials year, month and day for new ID full dataset
war_batting_people_teams['new_id_x'] = war_batting_people_teams['nameFirst_2'] + war_batting_people_teams['nameLast_2'] + war_batting_people_teams['birthYear_3'] #war_batting_people_teams['birthYear'].astype(str) #+ war_batting_people_teams['birthMonth'].astype(str) + war_batting_people_teams['birthDay'].astype(str) # + war_batting_people_teams['birthYear'].astype(str)



C:\Users\Hector G. Guerrero\AppData\Local\Temp\ipykernel_29804\3761326953.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  war_batting_people_teams_85['birthYear'] = war_batting_people_teams_85['birthYear'].astype('Int64')
C:\Users\Hector G. Guerrero\AppData\Local\Temp\ipykernel_29804\3761326953.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  war_batting_people_teams_85['birthMonth'] = war_batting_people_teams_85['birthMonth'].astype('Int64')
C:\Users\Hector G. Guerrero\AppData\Local\Temp\ipykernel_2

In [443]:
# Sanity check
show_player_info(war_batting_people_teams, 'Joey Gallo')[['name_common', 'year_ID', 'team_ID', 'birthYear', 'birthMonth', 'birthDay', 'new_id_x']]

name_common  year_ID team_ID  birthYear  birthMonth  birthDay new_id_x
56527  Joey Gallo     2015     TEX       1993          11        19  JoGa199
56528  Joey Gallo     2016     TEX       1993          11        19  JoGa199
56529  Joey Gallo     2017     TEX       1993          11        19  JoGa199
56530  Joey Gallo     2018     TEX       1993          11        19  JoGa199
56531  Joey Gallo     2019     TEX       1993          11        19  JoGa199
56532  Joey Gallo     2020     TEX       1993          11        19  JoGa199
56533  Joey Gallo     2021     TEX       1993          11        19  JoGa199
56534  Joey Gallo     2022     NYY       1993          11        19  JoGa199

__Unique ID for salary_pre__

In [444]:
# Extract year, month and day from borndate
salaries_pre['b_year'] = salaries_pre['borndate'].dt.year.astype('Int64')
salaries_pre['b_month'] = salaries_pre['borndate'].dt.month.astype('Int64')
salaries_pre['b_day'] = salaries_pre['borndate'].dt.day.astype('Int64')
salaries_pre['b_year_3'] = salaries_pre['b_year'].astype(str).str[:3]

# First and second letter of firstname and lastname
salaries_pre['firstname_2'] = salaries_pre['firstname'].str[:2]
salaries_pre['lastname_2'] = salaries_pre['lastname'].str[:2]

# concat initials year, month and day for new ID
salaries_pre['new_id_y'] = salaries_pre['firstname_2'] + salaries_pre['lastname_2'] + salaries_pre['b_year_3'] #salaries_pre['b_year'].astype(str) #+ salaries_pre['b_month'].astype(str) + salaries_pre['b_day'].astype(str) # + salaries_pre['b_year'].astype(str)


In [445]:
# Sanity check
# Joey Gallo
salaries_pre[salaries_pre['name'] == 'Joey Gallo'][['name', 'firstname_2', 'lastname_2', 'borndate', 'b_year', 'b_month', 'b_day', 'new_id_y']]

name firstname_2 lastname_2   borndate  b_year  b_month  b_day  \
14141  Joey Gallo          Jo         Ga 1993-11-19    1993       11     19   
14142  Joey Gallo          Jo         Ga 1993-11-19    1993       11     19   
14143  Joey Gallo          Jo         Ga 1993-11-19    1993       11     19   
14144  Joey Gallo          Jo         Ga 1993-11-19    1993       11     19   
14145  Joey Gallo          Jo         Ga 1993-11-19    1993       11     19   
14146  Joey Gallo          Jo         Ga 1993-11-19    1993       11     19   
14147  Joey Gallo          Jo         Ga 1993-11-19    1993       11     19   
14148  Joey Gallo          Jo         Ga 1993-11-19    1993       11     19   

      new_id_y  
14141  JoGa199  
14142  JoGa199  
14143  JoGa199  
14144  JoGa199  
14145  JoGa199  
14146  JoGa199  
14147  JoGa199  
14148  JoGa199

In [446]:
war_batting_people_teams.shape

(103404, 99)

In [447]:
# Join war_batting_people_teams_85 and slaries_pre on new_id_x = new_id_y and year_ID = year
war_batting_people_teams_salaries = war_batting_people_teams.merge(salaries_pre, how='left', left_on=['new_id_x', 'year_ID'], right_on=['new_id_y', 'year'])

In [448]:
# null values
war_batting_people_teams_salaries.isnull().sum()

name_common           0
age_x                 0
year_ID               0
player_ID             0
pitcher               0
mlb_ID                0
stint_ID              0
PA                    0
G_x                   0
Inn                   0
salary_x              0
OPS_plus              0
WAR                   0
WAR_def               0
WAR_off               0
team_ID               0
playerID            775
birthYear           775
birthMonth         1094
birthDay           1415
birthCountry        777
nameFirst           775
nameLast            775
weight             1689
height             1618
bats               2393
throws             1996
debut               814
finalGame           814
retroID             795
bbrefID             775
yearID              817
G_y                 817
AB_x                817
R_x                 817
H_x                 817
2B_x                817
3B_x                817
HR_x                817
RBI                 817
SB_x                817
CS_x            

In [449]:
# shape
war_batting_people_teams_salaries.shape

(112662, 120)

In [450]:
# rows with null values and year_ID = 2022
war_batting_people_teams_salaries[(war_batting_people_teams_salaries['year_ID'] == 2022) & (war_batting_people_teams_salaries['mlbid'].isnull())].head()




name_common  age_x  year_ID  player_ID pitcher     mlb_ID  stint_ID  \
4            AJ Alexy   24.0     2022  alexyaj01       Y   669935.0         1   
52          AJ Ladwig   29.0     2022  ladwiaj01       Y   607609.0         1   
345     Aaron Sanchez   29.0     2022  sanchaa01       Y  1185434.0         2   
395  Aaron Whitefield   25.0     2022  whiteaa01       N   664334.0         1   
441   Abraham Almonte   33.0     2022  almonab01       N   501659.0         1   

       PA  G_x   Inn  salary_x    OPS_plus   WAR  WAR_def  WAR_off team_ID  \
4     0.0    0   7.0       0.0    0.000000  0.00     0.00     0.00     TEX   
52    0.0    0   3.3       0.0    0.000000  0.00     0.00     0.00     MIA   
345   0.0    0  60.0  100000.0    0.000000  0.00     0.00     0.00     WSN   
395  11.0    5  24.7       0.0 -100.000000 -0.30    -0.01    -0.30     LAA   
441  37.0   15  83.0       0.0   91.078576 -0.03    -0.24     0.18     BOS   

      playerID  birthYear  birthMonth  birthDay birthCountry nameFirst  \
4    alexyaj01       1998           4        21          USA        AJ   
52   ladwiaj01       1992          12        24          USA        AJ   
345  sanchaa01       1992           7         1          USA     Aaron   
395  whiteaa01       1996           9         2    Australia     Aaron   
441  almonab01       1989           6        27         D.R.   Abraham   

       nameLast  weight  height bats throws       debut   finalGame   retroID  \
4         Alexy   195.0    76.0    R      R  2021-08-30  2022-09-12  alexa001   
52       Ladwig   220.0    77.0    R      R  2022-08-13  2022-08-13  ladwa001   
345     Sanchez   212.0    76.0    R      R  2014-07-23  2022-10-04  sanca006   
395  Whitefield   210.0    76.0    R      R  2020-07-25  2022-05-14  whita002   
441     Almonte   223.0    70.0    B      R  2013-08-30  2022-10-05  almoa001   

       bbrefID  yearID   G_y  AB_x  R_x  H_x  2B_x  3B_x  HR_x  RBI  SB_x  \
4    alexyaj01  2022.0   4.0   0.0  0.0  0.0   0.0   0.0   0.0  0.0   0.0   
52   ladwiaj01  2022.0   1.0   0.0  0.0  0.0   0.0   0.0   0.0  0.0   0.0   
345  sanchaa01  2022.0  15.0   0.0  0.0  0.0   0.0   0.0   0.0  0.0   0.0   
395  whiteaa01  2022.0   5.0  11.0  0.0  0.0   0.0   0.0   0.0  0.0   0.0   
441  almonab01  2022.0  15.0  35.0  7.0  9.0   2.0   0.0   1.0  2.0   1.0   

     CS_x  BB_x  SO_x  IBB  HBP_x   SH  SF_x  GIDP teamID lgID franchID divID  \
4     0.0   0.0   0.0  0.0    0.0  0.0   0.0   0.0    TEX   AL      TEX     W   
52    0.0   0.0   0.0  0.0    0.0  0.0   0.0   0.0    MIA   NL      FLA     E   
345   0.0   0.0   0.0  0.0    0.0  0.0   0.0   0.0    WAS   NL      WSN     E   
395   0.0   0.0   5.0  0.0    0.0  0.0   0.0   0.0    LAA   AL      ANA     W   
441   0.0   1.0  12.0  0.0    1.0  0.0   0.0   1.0    BOS   AL      BOS     E   

     Rank      G  Ghome     W      L DivWin LgWin WSWin    R_y    AB_y  \
4     4.0  162.0   81.0  68.0   94.0      N     N     N  707.0  5478.0   
52    4.0  162.0   81.0  69.0   93.0      N     N     N  586.0  5395.0   
345   5.0  162.0   81.0  55.0  107.0      N     N     N  603.0  5434.0   
395   3.0  162.0   81.0  73.0   89.0      N     N     N  623.0  5423.0   
441   5.0  162.0   81.0  78.0   84.0      N     N     N  735.0  5539.0   

        H_y   2B_y  3B_y   HR_y   BB_y    SO_y   SB_y  CS_y  HBP_y  SF_y  \
4    1308.0  224.0  20.0  198.0  456.0  1446.0  128.0  41.0   47.0  38.0   
52   1241.0  248.0  20.0  144.0  436.0  1429.0  122.0  29.0   70.0  36.0   
345  1351.0  252.0  20.0  136.0  442.0  1221.0   75.0  31.0   60.0  37.0   
395  1265.0  219.0  31.0  190.0  449.0  1539.0   77.0  27.0   54.0  25.0   
441  1427.0  352.0  12.0  155.0  478.0  1373.0   52.0  20.0   63.0  50.0   

        RA     ER   ERA   CG   SHO    SV  IPouts      HA    HRA    BBA  \
4    743.0  673.0  4.22  1.0  10.0  37.0  4305.0  1345.0  169.0  581.0   
52   676.0  617.0  3.86  6.0  10.0  41.0  4312.0  1311.0  173.0  511.0   
345  855.0  785.0  5.00  2.0   4.0  28.0  4235.0

---

In [451]:
war_batting_people_teams_salaries = war_batting_people_teams_salaries[~((war_batting_people_teams_salaries['year_ID'] == 2022) & (war_batting_people_teams_salaries['mlbid'].isnull()))]


---

In [453]:
pd.set_option('display.max_rows', None)
# null values 
war_batting_people_teams_salaries.isnull().sum().sort_values(ascending=False)

new_id_y          57855
averagesalary     57855
firstname         57855
lastname          57855
playerid          57855
mlbid             57855
year              57855
salary_y          57855
TeamName          57855
age_y             57855
teamrank          57855
leaguerank        57855
leagueminimum     57855
borndate          57855
name_y            57855
b_year            57855
b_month           57855
b_day             57855
b_year_3          57855
firstname_2       57855
lastname_2        57855
DivWin            45131
SF_y              44715
divID             43882
HBP_y             35788
CS_y              22314
WSWin              8498
Ghome              8489
attendance         5744
SB_y               2906
bats               2393
LgWin              2066
throws             1996
weight             1689
height             1618
park               1540
lgID               1458
birthDay           1415
SO_y               1409
birthMonth         1094
HR_y                817
BB_y            

__Some pre cleaning (pre_salaries)__

Rows that we realized during this step that contain null values and dropping them would not affect the analysis, as they are not relevant players with relevant stats.

In [457]:
drop_rows = ['ChJi197', 
             'ChTr196', 
             'DuSi197', 
             'FeLu197', 
             'FrMa196', 
             'GiRe196', 
             'JiPa195', 
             'KeMo197', 
             'MaDa196', 
             'NaCa198',
             'RaCa196',
             'WaMc197',
             ]

# Drop rows woth new_id_y in drop_rows from salaries_pre
salaries_pre = salaries_pre[~salaries_pre['new_id_y'].isin(drop_rows)]

In [458]:
special_drop_dict_salaries = {'JaLa197': [2010],
                     'LoTh196': [1985],
                     }

In [459]:
for key, value in special_drop_dict_salaries.items():
    salaries_pre  = salaries_pre[~((salaries_pre['new_id_y'] == key) & (salaries_pre['year'].isin(value)))]

##### Dropping newly created columns

Now that we succesfully joined the datasets, we can drop the columns (new ids) we created to join them for cleanliness purposes.

In [467]:
new_ids = ['birthYear_3', 'nameFirst_2', 'nameLast_2', 'new_id_x', 'firstname_2', 'lastname_2', 'new_id_y', 'b_year_3', 'b_year', 'b_month', 'b_day', 'birthMonth', 'birthDay', 'nameFirst', 'nameLast']

# Drop new_ids columns
war_batting_people_teams_salaries.drop(new_ids, axis=1, inplace=True)

KeyError: "['birthYear_3', 'nameFirst_2', 'nameLast_2', 'new_id_x', 'firstname_2', 'lastname_2', 'new_id_y', 'b_year_3', 'b_year', 'b_month', 'b_day'] not found in axis"

In [464]:
# sort by name_common and year_ID
war_batting_people_teams_salaries = war_batting_people_teams_salaries.sort_values(by=['year_ID', 'name_common'])

In [465]:
war_batting_people_teams_salaries.head()

name_common  age_x  year_ID  player_ID pitcher    mlb_ID  stint_ID  \
1065     Al Barker   32.0     1871  barkeal01       N  110565.0         1   
1682      Al Pratt   23.0     1871  prattal01       Y  120742.0         1   
1685      Al Reach   31.0     1871  reachal01       N  120965.0         1   
1800   Al Spalding   20.0     1871  spaldal01       Y  122558.0         1   
3906  Andy Leonard   25.0     1871  leonaan01       N  117686.0         1   

         PA  G_x  Inn  salary_x    OPS_plus   WAR  WAR_def  WAR_off team_ID  \
1065    5.0    1  0.0       0.0   92.724026  0.03     0.00     0.03     ROK   
1682  131.0   29  0.0       0.0   97.798322  0.01    -0.09     0.08     CLE   
1685  138.0   26  0.0       0.0  145.954197  0.98     0.02     0.99     ATH   
1800  152.0   31  0.0    1500.0   90.566468 -0.08    -0.15     0.05     BOS   
3906  151.0   31  0.0       0.0   99.733132  0.49    -0.08     0.56     OLY   

       playerID  birthYear  birthMonth  birthDay    birthCountry nameFirst  \
1065  barkeal01       1839           1        18             USA        Al   
1682  prattal01       1847          11        19             USA        Al   
1685  reachal01       1840           5        25  United Kingdom        Al   
1800  spaldal01       1850           9         2             USA        Al   
3906  leonaan01       1846           6         1         Ireland      Andy   

      nameLast  weight  height bats throws       debut   finalGame   retroID  \
1065    Barker   162.0    72.0  NaN    NaN  1871-06-01  1871-06-01  barka101   
1682     Pratt   140.0    67.0  NaN      R  1871-05-04  1872-08-19  prata101   
1685     Reach   155.0    66.0    L      L  1871-05-20  1875-05-21  reaca101   
1800  Spalding   170.0    73.0    R      R  1871-05-05  1878-08-31  spala101   
3906   Leonard   168.0    67.0    R      R  1871-05-05  1880-07-06  leona101   

        bbrefID  yearID   G_y   AB_x   R_x   H_x  2B_x  3B_x  HR_x   RBI  \
1065  barkeal01  1871.0   1.0    4.0   0.0   1.0   0.0   0.0   0.0   2.0   
1682  prattal01  1871.0  29.0  130.0  31.0  34.0   6.0   8.0   0.0  20.0   
1685  reachal01  1871.0  26.0  133.0  43.0  47.0   7.0   6.0   0.0  34.0   
1800  spaldal01  1871.0  31.0  144.0  43.0  39.0  10.0   1.0   1.0  31.0   
3906  leonaan01  1871.0  31.0  148.0  33.0  43.0   8.0   3.0   0.0  30.0   

      SB_x  CS_x  BB_x  SO_x  IBB  HBP_x   SH  SF_x  GIDP teamID lgID  \
1065   0.0   0.0   1.0   0.0  0.0    0.0  0.0   0.0   0.0    RC1  NaN   
1682   1.0   0.0   1.0   0.0  0.0    0.0  0.0   0.0   2.0    CL1  NaN   
1685   2.0   0.0   5.0   6.0  0.0    0.0  0.0   0.0   0.0    PH1  NaN   
1800   2.0   0.0   8.0   1.0  0.0    0.0  0.0   0.0   0.0    BS1  NaN   
3906  14.0   3.0   3.0   1.0  0.0    0.0  0.0   0.0   2.0    WS3  NaN   

     franchID divID  Rank     G  Ghome     W     L DivWin LgWin WSWin    R_y  \
1065      ROK   NaN   9.0  25.0    NaN   4.0  21.0    NaN     N   NaN  231.0   
1682      CFC   NaN   8.0  29.0    NaN  10.0  19.0    NaN     N   NaN  249.0   
1685      PNA   NaN   1.0  28.0    NaN  21.0   7.0    NaN     Y   NaN  376.0   
1800      BNA   NaN   3.0  31.0    NaN  20.0  10.0    NaN     N   NaN  401.0   
3906      OLY   NaN   4.0  32.0    NaN  15.0  15.0    NaN     N   NaN  310.0   

        AB_y    H_y  2B_y  3B_y  HR_y  BB_y  SO_y  SB_y  CS_y  HBP_y  SF_y  \
1065  1036.0  274.0  44.0  25.0   3.0  38.0  30.0  53.0  10.0    NaN   NaN   
1682  1186.0  328.0  35.0  40.0   7.0  26.0  25.0  18.0   8.0    NaN   NaN   
1685  1281.0  410.0  66.0  27.0   9.0  46.0  23.0  56.0  12.0    NaN   NaN   
1800  1372.0  426.0  70.0  37.0   3.0  60.0  19.0  73.0  16.0    NaN   NaN   
3906  1353.0  375.0  54.0  26.0   6.0  48.0  13.0  48.0  13.0    NaN   NaN   

         RA     ER   ERA    CG  SHO   SV  IPouts     HA   HRA   BBA   SOA  \
1065  287.0  108.0  4.30  23.0  1.0  0.0   678.0  315.0   3.0  34.0  16.0   
1682  341.0  116.0  4.11  23.0  0.0  0.0   762.0  346.0  13.0  53.0  34.0   
1685  266.0  137.0  4.95  27.0  0.0  0.0  

In [466]:
# Copy and save to csv
war_batting_people_teams_salaries_pre = war_batting_people_teams_salaries.copy()
war_batting_people_teams_salaries_pre.to_csv('pre_datasets/war_batting_people_teams_salaries_pre.csv', index=False)

---

---

### Next Steps <a class="anchor" id="summary"></a>
- Data wrangling on joined dataset to clean and prepare data for modeling
- EDA on joined dataset to identify potential features for modeling


-----
-----
-----